// rate-limit-memcached/lib/memcached-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Memcached = require("memcached");
class MemcachedStore {
    constructor(options) {
        const minutes = 15;
        const seconds = 60;
        this.expiration = options !== undefined && options.expiration !== undefined ? options.expiration : minutes * seconds;
        this.prefix = options !== undefined && options.prefix !== undefined ? options.prefix : 'rl:';
        this.client = options !== undefined && options.client !== undefined ? options.client : new Memcached(['127.0.0.1:11211']);
    }
    incr(key, cb) {
        this.client.increment(`${this.prefix}${key}`, 1, (err, result) => {
            if (err) {
                return cb(err, null);
            }
            if (result === false) {
                this.client.set(`${this.prefix}${key}`, 1, this.expiration, (ng, ok) => {
                    if (ok === true) {
                        cb(null, 1);
                    }
                    else {
                        cb(ng, null);
                    }
                });
            }
            else {
                cb(null, result);
            }
        });
    }
    decrement(key) {
        this.client.decrement(`${this.prefix}${key}`, 1, (err, result) => {
            return { err, result };
        });
    }
    resetKey(key) {
        this.client.del(`${this.prefix}${key}`);
    }
}
exports.default = MemcachedStore;
